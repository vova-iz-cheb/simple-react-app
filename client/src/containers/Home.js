// переписал с хуков, чтобы протестировать с jest enzyme
import React from 'react';
import { NewsList } from './NewsList';
import { LoadBox } from '../components/LoadBox';
import {
  fetchingAllNews,
  fetchingAllNewsFailed,
  fetchingAllNewsSuccess,
} from '../actions/newsActions';
import { connect } from 'react-redux'; // добавлено

export class Home extends React.Component {
  componentDidMount() {
    const dispatch = this.props.dispatch;

    document.title = 'Simple React App';

    this.props.getNews(dispatch);
  }

  render() {
    const { dispatch, news, error, isFetching } = this.props;
    return (
      <>
        <h1>News:</h1>
        {error && (
          <div className="error">
            Ошибка загрузки, попробовать еще раз{' '}
            <button onClick={() => this.props.getNews(dispatch)}>again</button>
          </div>
        )}
        {isFetching && <LoadBox />}
        {!isFetching && !error && <NewsList news={news} />}
      </>
    );
  }
}

const mapStateToProps = store => {
  return {
    news: store.allNews.news,
    error: store.allNews.error,
    isFetching: store.allNews.isFetching,
    getNews: dispatch => {
      dispatch(fetchingAllNews);

      return fetch('/api/news')
        .then(response => response.json())
        .then(result => dispatch(fetchingAllNewsSuccess(result)))
        .catch(err => dispatch(fetchingAllNewsFailed(true)));
    },
  };
};

export default connect(mapStateToProps)(Home);
