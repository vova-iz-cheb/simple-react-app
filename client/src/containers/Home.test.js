import React from 'react';
import { Home } from './Home';
import { shallow } from 'enzyme';

describe('Home container', () => {
  const props = {
    news: [],
    error: '',
    isFetching: false,
    dispatch: () => {},
    getNews: () => {},
  };

  describe('Home container initial', () => {
    const mockGetNews = jest.fn();

    const nextProps = {
      ...props,
      getNews: mockGetNews,
    };

    const homeContainer = shallow(<Home {...nextProps} />);

    it('Snapshot', () => {
      expect(homeContainer).toMatchSnapshot();
    });

    it('Render initial', () => {
      // console.log(homeContainer.debug());
      expect(homeContainer.find('h1').text()).toEqual('News:');
      expect(homeContainer.find('.error')).toHaveLength(0);
      expect(homeContainer.find('LoadBox')).toHaveLength(0);
      expect(homeContainer.find('NewsList').html()).toEqual('<p>К сожалению нет новостей.</p>');
    });

    it('dispatches the getNews() method it receives from props', () => {
      expect(mockGetNews).toHaveBeenCalledTimes(1);
    });
  });

  describe('Home container loading', () => {
    const nextProps = {
      ...props,
      isFetching: true,
    };

    const homeContainer = shallow(<Home {...nextProps} />);

    it('Snapshot', () => {
      expect(homeContainer).toMatchSnapshot();
    });

    it('Loading', () => {
      // console.log(homeContainer.debug());
      expect(homeContainer.find('h1').text()).toEqual('News:');
      expect(homeContainer.find('.error')).toHaveLength(0);
      expect(homeContainer.find('NewsList')).toHaveLength(0);
      expect(homeContainer.find('LoadBox')).toHaveLength(1);
      expect(homeContainer.find('LoadBox').html()).toEqual(
        '<div class="loading"><div class="snipper"></div></div>'
      );
    });
  });

  describe('Home container fetch data', () => {
    const nextProps = {
      ...props,
      news: [1],
    };

    const homeContainer = shallow(<Home {...nextProps} />);

    it('Snapshot', () => {
      expect(homeContainer).toMatchSnapshot();
    });

    it('Fecth data', () => {
      // console.log(homeContainer.debug());
      expect(homeContainer.find('h1').text()).toEqual('News:');
      expect(homeContainer.find('.error')).toHaveLength(0);
      expect(homeContainer.find('NewsList')).toHaveLength(1);
      expect(homeContainer.find('LoadBox')).toHaveLength(0);
    });
  });

  describe('Home container error', () => {
    const nextProps = {
      ...props,
      error: 'Ошибка',
    };

    const homeContainer = shallow(<Home {...nextProps} />);

    it('Snapshot', () => {
      expect(homeContainer).toMatchSnapshot();
    });

    it('Error', () => {
      // console.log(homeContainer.debug());
      expect(homeContainer.find('h1').text()).toEqual('News:');
      expect(homeContainer.find('.error')).toHaveLength(1);
      expect(homeContainer.find('NewsList')).toHaveLength(0);
      expect(homeContainer.find('LoadBox')).toHaveLength(0);
    });
  });

  describe('Home container loading after error', () => {
    const nextProps = {
      ...props,
      error: 'Ошибка',
      isFetching: true,
    };

    const homeContainer = shallow(<Home {...nextProps} />);

    it('Snapshot', () => {
      expect(homeContainer).toMatchSnapshot();
    });

    it('Loading after error', () => {
      // console.log(homeContainer.debug());
      expect(homeContainer.find('h1').text()).toEqual('News:');
      expect(homeContainer.find('.error')).toHaveLength(1);
      expect(homeContainer.find('NewsList')).toHaveLength(0);
      expect(homeContainer.find('LoadBox')).toHaveLength(1);
    });
  });
});
