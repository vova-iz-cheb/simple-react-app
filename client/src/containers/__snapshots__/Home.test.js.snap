// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home container Home container error Snapshot 1`] = `
<Fragment>
  <h1>
    News:
  </h1>
  <div
    className="error"
  >
    Ошибка загрузки, попробовать еще раз
     
    <button
      onClick={[Function]}
    >
      again
    </button>
  </div>
</Fragment>
`;

exports[`Home container Home container fetch data Snapshot 1`] = `
<Fragment>
  <h1>
    News:
  </h1>
  <NewsList
    news={
      Array [
        1,
      ]
    }
  />
</Fragment>
`;

exports[`Home container Home container initial Snapshot 1`] = `
<Fragment>
  <h1>
    News:
  </h1>
  <NewsList
    news={Array []}
  />
</Fragment>
`;

exports[`Home container Home container loading Snapshot 1`] = `
<Fragment>
  <h1>
    News:
  </h1>
  <LoadBox />
</Fragment>
`;

exports[`Home container Home container loading after error Snapshot 1`] = `
<Fragment>
  <h1>
    News:
  </h1>
  <div
    className="error"
  >
    Ошибка загрузки, попробовать еще раз
     
    <button
      onClick={[Function]}
    >
      again
    </button>
  </div>
  <LoadBox />
</Fragment>
`;
